File: scala/src/MergeSort.scala
/**
 * Implementation of merge sort algorithm in Scala
 * Time complexity: O(n log n)
 * Space complexity: O(n)
 */
object MergeSort {
  /**
   * Sorts an array using merge sort algorithm.
   * @param arr The array to sort
   * @return The sorted array
   */
  def mergeSort[T: Ordering](arr: Array[T]): Array[T] = {
    if (arr.length <= 1) arr
    else {
      val mid = arr.length / 2
      val left = arr.slice(0, mid)
      val right = arr.slice(mid, arr.length)
      merge(mergeSort(left), mergeSort(right))
    }
  }

  /**
   * Merges two sorted arrays into a single sorted array.
   * @param left The left sorted array
   * @param right The right sorted array
   * @return The merged sorted array
   */
  private def merge[T: Ordering](left: Array[T], right: Array[T]): Array[T] = {
    val result = new Array[T](left.length + right.length)
    var i = 0
    var j = 0
    var k = 0
    val ord = implicitly[Ordering[T]]

    while (i < left.length && j < right.length) {
      if (ord.lteq(left(i), right(j))) {
        result(k) = left(i)
        i += 1
      } else {
        result(k) = right(j)
        j += 1
      }
      k += 1
    }

    while (i < left.length) {
      result(k) = left(i)
      i += 1
      k += 1
    }

    while (j < right.length) {
      result(k) = right(j)
      j += 1
      k += 1
    }

    result
  }

  /**
   * Checks if an array is sorted in ascending order.
   * @param arr The array to check
   * @return true if the array is sorted, false otherwise
   */
  def isSorted[T: Ordering](arr: Array[T]): Boolean = {
    val ord = implicitly[Ordering[T]]
    arr.sliding(2).forall { case Array(a, b) => ord.lteq(a, b) }
  }
}

/**
 * Main object for command-line interface and testing
 */
object Main {
  def main(args: Array[String]): Unit = {
    if (args.isEmpty) {
      println("Please provide an input file path")
      return
    }

    val file = args(0)
    val numbers = scala.io.Source.fromFile(file).getLines().map(_.toInt).toArray
    val sorted = MergeSort.mergeSort(numbers)
    java.nio.file.Files.write(
      java.nio.file.Paths.get("sorted_data.txt"),
      sorted.mkString("\n").getBytes
    )
  }
}

/**
 * Test suite
 */
object Tests {
  def runTests(): Boolean = {
    val testCases = List(
      "Empty array" -> (MergeSort.mergeSort(Array.empty[Int]).isEmpty),
      "Single element" -> (MergeSort.mergeSort(Array(1)).sameElements(Array(1))),
      "Two elements" -> (MergeSort.mergeSort(Array(2, 1)).sameElements(Array(1, 2))),
      "Multiple elements" -> (
        MergeSort.mergeSort(Array(5, 3, 8, 4, 2))
          .sameElements(Array(2, 3, 4, 5, 8))
      ),
      "Duplicate elements" -> (
        MergeSort.mergeSort(Array(3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5))
          .sameElements(Array(1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9))
      ),
      "Negative numbers" -> (
        MergeSort.mergeSort(Array(-3, 1, -4, 1, -5, 9, -2, 6, -5, 3, -5))
          .sameElements(Array(-5, -5, -5, -4, -3, -2, 1, 1, 3, 6, 9))
      ),
      "Large array" -> {
        val largeArray = (1 to 1000).toArray
        MergeSort.isSorted(MergeSort.mergeSort(largeArray))
      },
      "Strings" -> (
        MergeSort.mergeSort(Array("banana", "apple", "cherry"))
          .sameElements(Array("apple", "banana", "cherry"))
      )
    )

    println("Running tests...")
    var allPassed = true

    testCases.foreach { case (name, test) =>
      val passed = test
      println(s"$name: ${if (passed) "PASSED" else "FAILED"}")
      if (!passed) allPassed = false
    }

    allPassed
  }
}

// Run tests if DEBUG is set
object TestRunner {
  def main(args: Array[String]): Unit = {
    if (sys.env.getOrElse("DEBUG", "0") == "1") {
      if (Tests.runTests()) {
        println("All tests passed!")
      } else {
        println("Some tests failed!")
      }
    } else {
      Main.main(args)
    }
  }
}

File: zig/src/merge_sort.zig
const std = @import("std");

/// Merge sort implementation in Zig
/// Time complexity: O(n log n)
/// Space complexity: O(n)
///
/// @param allocator Memory allocator for temporary arrays
/// @param arr The array to sort
/// @return void (sorts in place)
pub fn mergeSort(allocator: std.mem.Allocator, arr: []i32) !void {
    if (arr.len <= 1) return;

    // Split array into two halves
    const mid = arr.len / 2;
    const left = arr[0..mid];
    const right = arr[mid..];

    // Recursively sort both halves
    try mergeSort(allocator, left);
    try mergeSort(allocator, right);

    // Merge the sorted halves
    try merge(allocator, arr, left, right);
}

/// Helper function to merge two sorted arrays
///
/// @param allocator Memory allocator for temporary array
/// @param arr The array to merge into
/// @param left First sorted array
/// @param right Second sorted array
/// @return void (merges in place)
fn merge(allocator: std.mem.Allocator, arr: []i32, left: []i32, right: []i32) !void {
    const left_size = left.len;
    const right_size = right.len;

    // Create temporary arrays
    var temp_left: []i32 = undefined;
    var temp_right: []i32 = undefined;

    // Copy data to temporary arrays
    temp_left = try allocator.alloc(i32, left_size);
    temp_right = try allocator.alloc(i32, right_size);
    defer allocator.free(temp_left);
    defer allocator.free(temp_right);

    for (0..left_size) |i| {
        temp_left[i] = left[i];
    }
    for (0..right_size) |i| {
        temp_right[i] = right[i];
    }

    // Merge temporary arrays back into arr
    var i: usize = 0;
    var j: usize = 0;
    var k: usize = 0;

    while (i < left_size and j < right_size) {
        if (temp_left[i] <= temp_right[j]) {
            arr[k] = temp_left[i];
            i += 1;
        } else {
            arr[k] = temp_right[j];
            j += 1;
        }
        k += 1;
    }

    // Copy remaining elements of temp_left if any
    while (i < left_size) {
        arr[k] = temp_left[i];
        i += 1;
        k += 1;
    }

    // Copy remaining elements of temp_right if any
    while (j < right_size) {
        arr[k] = temp_right[j];
        j += 1;
        k += 1;
    }
}

pub fn main() !void {
    // Get command line arguments
    var args = try std.process.argsWithAllocator(std.heap.page_allocator);
    defer args.deinit();

    // Skip program name
    _ = args.skip();

    // Get input file path
    const input_file = args.next() orelse {
        std.debug.print("Please provide a file path as an argument\n", .{});
        std.process.exit(1);
    };

    // Create a general purpose allocator
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read numbers from file
    const file = try std.fs.cwd().openFile(input_file, .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    var in_stream = buf_reader.reader();

    var numbers = std.ArrayList(i32).init(allocator);
    defer numbers.deinit();

    var buf: [1024]u8 = undefined;
    while (try in_stream.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        const num = try std.fmt.parseInt(i32, line, 10);
        try numbers.append(num);
    }

    // Sort the array
    try mergeSort(allocator, numbers.items);

    // Write sorted numbers back to file
    const out_file = try std.fs.cwd().createFile("sorted_data.txt", .{});
    defer out_file.close();

    var out_stream = out_file.writer();
    for (numbers.items) |num| {
        try out_stream.print("{d}\n", .{num});
    }
}
